08-rails/testing-controllers.md

```diff
@@ -1,178 +1,226 @@
+**Question:**  What is one example of a controller action that commonly redirects the user?
+
+
+Last is a way to ensure that the controller action appropriately changes the related model.  We'll see how to use the `must_change` matcher with some examples later on.
+
 ```ruby
-assert_difference('Post.count', -1) do
-  # Controller here that would delete something
-end
+proc {
+  # run the delete verb on the post_path with a param equal to 1
+  delete post_path(1)
+}.must_change 'Post.count', -1
 ```
 
+We still have the existing Minitest matchers, like `must_equal` and `must_be` to use as well.  
+
+
 ### Test Setup
-First, we need to execute the controller action that we would like to test. We use the verb along with the controller action to execute the controller.
+First, we need to execute the controller action that we would like to test. We use the verb along with the path to execute the controller method.
 
 ```ruby
-class StudentsControllerTest < ActionController::TestCase
-  test "should get the new form" do
-    get :new
+describe PostController do
+  it "should get index" do
+    get post_index_path
   end
 end
 ```
 
-Next, we need to use one (or more) of the assertions to ensure that the controller action executed as expected.
+The test starts with executing the controller with, in this case, a get request and a given path.  We could also put `get '/posts'` instead, assuming that our index page was at that path, but using the Rails path helper methods make it more portable and easier to change.  If the paths change the testing code doesn't need to be modified.  
+
+Next, we need to use one (or more) of the expectations to ensure that the controller action executed as expected.
+
 ```ruby
-class StudentsControllerTest < ActionController::TestCase
-  test "should get the new form" do
-    get :new
-    assert_response :success
-    assert_template :new
+describe PostController do
+  ...
+  it "should get index" do
+    get post_index_path
+    # Added expectation to verify the HTTP response code
+    must_respond_with :success
   end
 end
 ```
 
-Now you try it! Try setting up the next test for the `edit` action.
+**Exercise:** Now you try it! Try setting up the next test for the `new` action on your last project.
 
 ### Test Setup with Params
 When we create the controller actions, oftentimes they contain information that comes from the `params` hash, with data populated from the routes or forms. In order to appropriately test controllers, we must "mock" this information.
 
 Let's refresh our memory with a controller action:
 ```ruby
 def show
-  @student = Student.find(params[:id])
+  @post = Post.find(params[:id])
 end
 ```
 
-In this case, we need to provide a key of `id` with a corresponding value. We specify this in a hash to correspond to `params` along with the verb and action.
+In this case, we need to provide a key of `id` with a corresponding value in the path helper. 
```

Comment:
I think it would be more accurate to say that we pass the ID value as an argument to the path helper. Alternatively, we could change the code snippet below to say:
```ruby
get post_path(id: 1)
```
and then this line would be most accurate.